pypit/core/arextract.py

```diff
@@ -30,6 +30,934 @@
+    flux_opt = np.nansum(mask_sub*ivar_sub*img_sub*oprof_sub, axis=1)/(mivar_num + (mivar_num == 0.0))
+    # Optimally extracted noise variance (sky + read noise) only. Since
+    # this variance is not the same as that used for the weights, we
+    # don't get the usual cancellation. Additional denom factor is the
+    # analog of the numerator in Horne's variance formula. Note that we
+    # are only weighting by the profile (ivar_sub=1) because
+    # otherwise the result depends on the signal (bad).
+    nivar_num =np.nansum(mask_sub*oprof_sub**2, axis=1) # Uses unit weights
+    nvar_opt = ivar_denom*((mask_sub*vno_sub*oprof_sub**2).sum(axis=1))/(nivar_num**2 + (nivar_num**2 == 0.0))
+    nivar_opt = 1.0/(nvar_opt + (nvar_opt == 0.0))
+    # Optimally extract sky and (read noise)**2 in a similar way
+    sky_opt = ivar_denom*(np.nansum(mask_sub*sky_sub*oprof_sub**2, axis=1))/(nivar_num**2 + (nivar_num**2 == 0.0))
+    rn2_opt = ivar_denom*(np.nansum(mask_sub*rn2_sub*oprof_sub**2, axis=1))/(nivar_num**2 + (nivar_num**2 == 0.0))
+    rn_opt = np.sqrt(rn2_opt)
+    rn_opt[np.isnan(rn_opt)]=0.0
+
+    tot_weight = np.nansum(mask_sub*ivar_sub*oprof_sub, axis=1)
+    mask_opt = (tot_weight > 0.0) & (mivar_num > 0.0) & (ivar_denom > 0.0)
+    frac_use = np.nansum((mask_sub*ivar_sub > 0.0)*oprof_sub, axis=1)
+    # Use the same weights = oprof^2*mivar for the wavelenghts as the flux.
+    # Note that for the flux, one of the oprof factors cancels which does
+    # not for the wavelengths.
+    wave_opt = np.nansum(mask_sub*ivar_sub*wave_sub*oprof_sub**2, axis=1)/(mivar_num + (mivar_num == 0.0))
+    # Interpolate wavelengths over masked pixels
+    badwvs = (mivar_num <= 0) | (np.isfinite(wave_opt) == False) | (wave_opt <= 0.0)
+    if badwvs.any():
+        oprof_smash = np.nansum(oprof_sub**2, axis=1)
+        # Can we use the profile average wavelengths instead?
+        oprof_good = badwvs & (oprof_smash > 0.0)
+        if oprof_good.any():
+            wave_opt[oprof_good] = np.nansum(wave_sub[oprof_good,:]*oprof_sub[oprof_good,:]**2, axis=1)/np.nansum(oprof_sub[oprof_good,:]**2, axis=1)
+        oprof_bad = badwvs & ((oprof_smash <= 0.0) | (np.isfinite(oprof_smash) == False) | (wave_opt <= 0.0) | (np.isfinite(wave_opt) == False))
+        if oprof_bad.any():
+            # For pixels with completely bad profile values, interpolate from trace.
+            f_wave = RectBivariateSpline(spec_vec,spat_vec, waveimg)
+            wave_opt[oprof_bad] = f_wave(specobj.trace_spec[oprof_bad], specobj.trace_spat[oprof_bad],grid=False)
+
+    flux_model = np.outer(flux_opt,np.ones(nsub))*oprof_sub
+    chi2_num = np.nansum((img_sub - flux_model)**2*ivar_sub*mask_sub,axis=1)
+    chi2_denom = np.fmax(np.nansum(ivar_sub*mask_sub > 0.0, axis=1) - 1.0, 1.0)
+    chi2 = chi2_num/chi2_denom
+
+    # Fill in the optimally extraction tags
+    specobj.optimal['WAVE_OPT'] = wave_opt    # Optimally extracted wavelengths
```

Comment:
Add all of this to the doc string, i.e.
what is being filled up into the specobj Object.